telemetry/internal/appinsights/inmemorychannel.go

```diff
@@ -0,0 +1,453 @@
+package appinsights
+
+import (
+	"log"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/microsoft/go-infra/telemetry/internal/appinsights/internal/contracts"
+)
+
+var (
+	submit_retries = []time.Duration{time.Duration(10 * time.Second), time.Duration(30 * time.Second), time.Duration(60 * time.Second)}
+)
+
+// inMemoryChannel stores events exclusively in memory.
+// Presently the only telemetry channel implementation available.
+type inMemoryChannel struct {
+	endpointAddr  string
+	collectChan   chan *contracts.Envelope
+	controlChan   chan *inMemoryChannelControl
+	batchSize     int
+	batchInterval time.Duration
+	waitgroup     sync.WaitGroup
+	throttle      *throttleManager
+	transmitter   transmitter
+	errorLog      *log.Logger
+
+	closed atomic.Bool
+}
+
+type inMemoryChannelControl struct {
+	// If true, flush the buffer.
+	flush bool
+
+	// If true, stop listening on the channel.  (Flush is required if any events are to be sent)
+	stop bool
+
+	// If stopping and flushing, this specifies whether to retry submissions on error.
+	retry bool
+
+	// If retrying, what is the max time to wait before finishing up?
+	timeout time.Duration
+
+	// If specified, a message will be sent on this channel when all pending telemetry items have been submitted
+	callback chan struct{}
+}
+
+// newInMemoryChannel creates an inMemoryChannel instance and starts a background submission goroutine.
```

Comment:
Maybe mention what's necessary to make sure you don't leak a goroutine after getting a result from this func?

Reply:
Good point. I've moved the `acceptLoop` call outside of this constructor to let the caller control when the loop is started.